src/commands/play.rs

```diff
@@ -1,22 +1,156 @@
+use std::sync::Arc;
+
 use crate::{
-    strings::{AUTHOR_NOT_FOUND, MISSING_PLAY_QUERY, NO_VOICE_CONNECTION},
+    commands::{now_playing::now_playing, summon::summon, EnqueueType, PlayFlag},
+    strings::{MISSING_PLAY_QUERY, NO_VOICE_CONNECTION},
     utils::{get_human_readable_timestamp, send_simple_message},
 };
 
 use serenity::{
-    builder::CreateEmbedFooter,
     client::Context,
     framework::standard::{macros::command, Args, CommandResult},
     model::channel::Message,
+    prelude::Mutex,
 };
 
-use songbird::input::Restartable;
-
+use songbird::{input::Restartable, tracks::TrackHandle, Call};
 use youtube_dl::{YoutubeDl, YoutubeDlOutput};
 
-#[command]
-#[aliases("p")]
-async fn play(ctx: &Context, msg: &Message, mut args: Args) -> CommandResult {
+async fn enqueue_song(
+    call: &Arc<Mutex<Call>>,
+    query: String,
+    is_url: bool,
+    flag: &PlayFlag,
+) -> Vec<TrackHandle> {
+    let source = match is_url {
+        true => Restartable::ytdl(query, true).await.unwrap(),
+        false => Restartable::ytdl_search(query, false).await.unwrap(),
+    };
+
+    let mut handler = call.lock().await;
+    handler.enqueue_source(source.into());
+    let mut queue_snapshot = handler.queue().current_queue();
+    drop(handler);
+
+    match flag {
+        PlayFlag::PLAYTOP => {
+            if queue_snapshot.len() > 2 {
+                let handler = call.lock().await;
+
+                handler.queue().modify_queue(|queue| {
+                    let mut not_playing = queue.split_off(1);
+                    not_playing.rotate_right(1);
+                    queue.append(&mut not_playing);
+                });
+
+                queue_snapshot = handler.queue().current_queue();
+            }
+        }
+        _ => (),
+    }
+
+    queue_snapshot
+}
+
+async fn enqueue_playlist(
+    call: &Arc<Mutex<Call>>,
+    uri: &String,
+    flag: &PlayFlag,
+) -> Vec<TrackHandle> {
+    let mut entries = vec![];
+
+    match YoutubeDl::new(uri).flat_playlist(true).run() {
+        Ok(result) => {
+            if let YoutubeDlOutput::Playlist(playlist) = result {
+                entries = playlist.entries.unwrap();
+
+                for entry in &entries {
+                    let uri = format!(
+                        "https://www.youtube.com/watch?v={}",
+                        entry.url.as_ref().unwrap()
+                    );
+                    let source = Restartable::ytdl(uri, true).await.unwrap();
+                    let mut handler = call.lock().await;
+                    handler.enqueue_source(source.into());
+                }
```

Comment:
Can't you just use `enqueue_song` for each entry?

Reply:
Good thinking but I don't think so - at least with how the code is structured right now. Because `enqueue_song` places the song at the top of the queue, the playlist should be enqueued in reverse. I could reverse the list of songs in a playlist and pass it to `enqueue_song`. A little hacky but should eliminate the last match in this method, no?

Reply:
I think it makes sense to reuse `enqueue_song`. Calling reverse on the list of songs to enqueue is good enough. Alternatively, that can be a boolean on the enqueue song method like "reverse"

Reply:
Well, I didn't think this through because imagine playlist `[A,B,C,D]` and an empty queue.
If you `!playtop` the reversed playlist, you get `[D,A,B,C]` - D immediately starts playing and the rest get pushed under.

Do you think it is that useful to pushtop a playlist? I'm currently considering removing it to simplify the code.

Reply:
Can't you:

1- `[A, B, C, D]` => play(A) + enqueue_playlist([D, C, B])

or

2- `[A, B, C, D]` => enqueue_playlist([A, B, C, D], reverse=false)